components/scream/src/physics/shoc/shoc_diag_second_moments_impl.hpp

```diff
@@ -13,10 +13,68 @@ namespace shoc {
 
 template<typename S, typename D>
 KOKKOS_FUNCTION
-void Functions<S,D>::diag_second_moments(const Int& shcol, const Int& nlev, const Int& nlevi, const uview_1d<const Spack>& thetal, const uview_1d<const Spack>& qw, const uview_1d<const Spack>& u_wind, const uview_1d<const Spack>& v_wind, const uview_1d<const Spack>& tke, const uview_1d<const Spack>& isotropy, const uview_1d<const Spack>& tkh, const uview_1d<const Spack>& tk, const uview_1d<const Spack>& dz_zi, const uview_1d<const Spack>& zt_grid, const uview_1d<const Spack>& zi_grid, const uview_1d<const Spack>& shoc_mix, const uview_1d<Spack>& thl_sec, const uview_1d<Spack>& qw_sec, const uview_1d<Spack>& wthl_sec, const uview_1d<Spack>& wqw_sec, const uview_1d<Spack>& qwthl_sec, const uview_1d<Spack>& uw_sec, const uview_1d<Spack>& vw_sec, const uview_1d<Spack>& wtke_sec, const uview_1d<Spack>& w_sec)
+void Functions<S,D>::diag_second_moments(const MemberType& team, const Int& nlev, const Int& nlevi, 
+                     const uview_1d<const Spack>& thetal, const uview_1d<const Spack>& qw, const uview_1d<const Spack>& u_wind, 
+                     const uview_1d<const Spack>& v_wind, const uview_1d<const Spack>& tke, const uview_1d<const Spack>& isotropy, 
+                     const uview_1d<const Spack>& tkh, const uview_1d<const Spack>& tk, const uview_1d<const Spack>& dz_zi, 
+                     const uview_1d<const Spack>& zt_grid, const uview_1d<const Spack>& zi_grid, const uview_1d<const Spack>& shoc_mix,
+                     const uview_1d<Spack>& isotropy_zi, const uview_1d<Spack>& tkh_zi, const uview_1d<Spack>& tk_zi, 
+                     const uview_1d<Spack>& thl_sec, const uview_1d<Spack>& qw_sec, const uview_1d<Spack>& wthl_sec, const uview_1d<Spack>& wqw_sec, 
+                     const uview_1d<Spack>& qwthl_sec, const uview_1d<Spack>& uw_sec, const uview_1d<Spack>& vw_sec, const uview_1d<Spack>& wtke_sec, 
+                     const uview_1d<Spack>& w_sec)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  // Purpose of this subroutine is to diagnose the second
+  //  order moments needed for the SHOC parameterization.
+  //  Namely these are variances of thetal, qw, and vertical
+  //  velocity.  In addition the vertical fluxes of thetal, qw,
+  //  u, v, TKE, and tracers are computed here as well as the
+  //  correlation of qw and thetal.
+
+  const auto thl2tune = sp(1);
+
+  // moisture variance
+  const auto qw2tune = sp(1);
+
+  // temp moisture covariance
+  const auto qwthl2tune = sp(1);
+
+  // vertical velocity variance
+  const auto w2tune = sp(1);
+
+  // Interpolate some variables from the midpoint grid to the interface grid
+  linear_interp(team, zt_grid, zi_grid, isotropy, isotropy_zi, nlev, nlevi, sp(0));
+  linear_interp(team, zt_grid, zi_grid, tkh,      tkh_zi,      nlev, nlevi, sp(0));
+  linear_interp(team, zt_grid, zi_grid, tk,       tk_zi,       nlev, nlevi, sp(0)); 
```

Comment:
`sp()` is unnecessary above since `linear_interp` will do the correct conversion of integers -> Scalar. It doesn't hurt anything though, just may look nicer. Up to you.

Reply:
Also, 0 is exactly represented regardless of the numerical type.

Reply:
I would keep them as it is